libopenage/renderer/resources/parser/parse_sprite.cpp

```diff
@@ -0,0 +1,296 @@
+		std::vector<std::string> keywordargs{util::split(args[i], '=')};
+
+		if (unlikely(!keywordfuncs.contains(keywordargs[0]))) {
+			throw Error(MSG(err) << "Keyword argument "
+			                     << keywordargs[0]
+			                     << " of 'angle' attribute is not defined");
+		}
+
+		keywordfuncs[keywordargs[0]](keywordargs);
+	}
+
+	return angle;
+}
+
+/**
+ * Parse the frame attribute.
+ *
+ * @param args Arguments from the line with a \p frame attribute.
+ *             The first argument is expected to be the attribute keyword.
+ *
+ * @return Struct containing the attribute data.
+ */
+FrameData parse_frame(std::vector<std::string> args) {
+	FrameData frame;
+
+	frame.index = std::stoul(args[1]);
+	frame.angle = std::stoul(args[2]);
+	frame.layer_id = std::stoul(args[3]);
+	frame.texture_id = std::stoul(args[4]);
+	frame.subtex_id = std::stoul(args[5]);
+
+	return frame;
+}
+
+Animation2dInfo parse_sprite_file(const util::Path &file) {
+	if (unlikely(!file.is_file())) {
+		throw Error(MSG(err) << "Reading .sprite file '"
+		                     << file.get_name()
+		                     << "' failed. Reason: File not found");
+	}
+
+	auto content = file.open();
+	auto lines = content.get_lines();
+
+	float scalefactor = 1.0;
+	std::vector<TextureData> textures;
+	std::vector<LayerData> layers;
+	std::vector<AngleData> angles;
+
+	// Map frame data to angle
+	std::unordered_map<size_t, std::vector<FrameData>> frames;
+
+	const auto keywordfuncs = datastructure::create_const_map<std::string, std::function<void(std::vector<std::string>)>>(
+		std::make_pair("version", [&](std::vector<std::string> args) {
+			size_t version_no = parse_spriteversion(args);
+
+			if (version_no != 2) {
+				throw Error(MSG(err) << "Reading .sprite file '"
+			                         << file.get_name()
+			                         << "' failed. Reason: Version "
+			                         << version_no << " not supported");
+			}
+		}),
+		std::make_pair("texture", [&](std::vector<std::string> args) {
+			textures.push_back(parse_texture(args));
+		}),
+		std::make_pair("scalefactor", [&](std::vector<std::string> args) {
+			scalefactor = parse_scalefactor(args);
+		}),
+		std::make_pair("layer", [&](std::vector<std::string> args) {
+			layers.push_back(parse_layer(args));
+		}),
+		std::make_pair("angle", [&](std::vector<std::string> args) {
+			angles.push_back(parse_angle(args));
+		}),
+		std::make_pair("frame", [&](std::vector<std::string> args) {
+			auto frame = parse_frame(args);
+			if (frames.count(frame.angle) == 0) {
+				std::vector<FrameData> angle_frames{};
+				frames.emplace(std::make_pair(frame.angle, angle_frames));
+			}
+			frames.at(frame.angle).push_back(frame);
+		}));
+
+	for (auto line : lines) {
+		// Skip empty lines and comments
+		if (line.empty() || line.substr(0, 1) == "#") {
+			continue;
+		}
+		std::vector<std::string> args{util::split(line, ' ')};
+
+		if (unlikely(!keywordfuncs.contains(args[0]))) {
```

Comment:
I've decided to not do that right now because I know it will confuse me a lot :D I've placed TODOs so we can change it later when the code doesn't change so much anymore.